pkg/datasource/plugin.go

```diff
@@ -0,0 +1,201 @@
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"google.golang.org/api/iterator"
+)
+
+type queryResult struct {
+	Time   time.Time `bigquery:"time"`
+	Values int64     `bigquery:"metric"`
+}
+
+// TransformedResults contains the results from BigQuery for Grafana
+type TransformedResults struct {
+	Time   []time.Time
+	Values []int64
+}
+
+type queryModel struct {
+	Format string `json:"format"`
+	Dataset          string   `json:"dataset"`
+	Group            []string `json:"group"`
+	MetricColumn     string   `json:"metricColumn"`
+	OrderByCol       string   `json:"orderByCol"`
+	OrderBySort      string   `json:"orderBySort"`
+	Partitioned      bool     `json:"partitioned"`
+	PartitionedField string   `json:"partitionedField"`
+	Project          string   `json:"project"`
+	RawQuery         bool     `json:"rawQuery"`
+	RawSQL           string   `json:"rawSql"`
+	RefID            string   `json:"refId"`
+	Sharded          bool     `json:"sharded"`
+	Table            string   `json:"table"`
+	TimeColumn       string   `json:"timeColumn"`
+	TimeColumnType   string   `json:"timeColumnType"`
+	Location         string   `json:"location"`
+}
+
+type instanceSettings struct {
+	httpClient *http.Client
+}
+
+func newDataSourceInstance(_ backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
+	return &instanceSettings{
+		httpClient: &http.Client{},
+	}, nil
+}
+
+// Dispose is called before creating a new instance to allow plugin authors to cleanup
+func (s *instanceSettings) Dispose() {}
+
+// BigQueryDatasource is an example datasource used to scaffold
+// new datasource plugins with an backend.
+type BigQueryDatasource struct {
+	log.Logger
+	im instancemgmt.InstanceManager
+}
+
+// New creates a datasource that is sued for querying BigQuery
+func New(logger log.Logger) datasource.ServeOpts {
+	im := datasource.NewInstanceManager(newDataSourceInstance)
+	ds := &BigQueryDatasource{
+		Logger: logger,
+		im: im,
+	}
+
+	return datasource.ServeOpts{
+		QueryDataHandler:   ds,
+		CheckHealthHandler: ds,
+	}
+}
+
+// QueryData handles multiple queries and returns multiple responses.
+// req contains the queries []DataQuery (where each query contains RefID as a unique identifier).
+// The QueryDataResponse contains a map of RefID to the response for each query, and each response
+// contains Frames ([]*Frame).
+func (bq *BigQueryDatasource) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	response := backend.NewQueryDataResponse()
+	mtx := &sync.Mutex{}
+
+	ctx, cancel := context.WithTimeout(ctx, time.Minute)
+	defer cancel()
+	// Execute the queries in parallel and collect the responses.
+	for _, q := range req.Queries {
+		go func(dq backend.DataQuery) {
+			res := bq.doQuery(ctx, dq)
+			if !errors.Is(ctx.Err(), context.DeadlineExceeded) {
+				mtx.Lock()
+				response.Responses[dq.RefID] = res
+				mtx.Unlock()
+			}
+		}(q)
+	}
+
+	return response, nil
```

Comment:
We need to wait for the results. I'd typically use a wait group here.

This is also a place where we might want to control concurrency - limiting the maximum count of in-flight requests to BQ.

Reply:
So, funny story. I did this on Friday (waitgroup + concurrency control). It turns out this code is not actually used anywhere. The requests to BigQuery are done in the TypeScript. This code is a lie.

What I did, and then 20 minutes of trying to figure out why the Log line never got hit.
```
const maxParallelQueries = 20

func (bq *BigQueryDatasource) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
	response := backend.NewQueryDataResponse()
	mtx := &sync.Mutex{}
	wg := &sync.WaitGroup{}
	queries := req.Queries

	log.DefaultLogger.Info("Running %d queries", len(queries))
	// Execute the queries in parallel, up to maxParallelQueries, and collect the responses.
	for {
		var next = maxParallelQueries
		if llen := len(queries); llen < maxParallelQueries {
			next = llen
		}
		log.DefaultLogger.Warn("Running %d queries", next)

		wg.Add(next)
		qctx, cancel := context.WithTimeout(ctx, time.Minute)
		for _, q := range queries[:next] {
			go func(dq backend.DataQuery) {
				//res := bq.doQuery(qctx, dq)
				if !errors.Is(qctx.Err(), context.DeadlineExceeded) {
					mtx.Lock()
					response.Responses[dq.RefID] = backend.DataResponse{}
					mtx.Unlock()
				}
				wg.Done()
			}(q)
		}
		wg.Wait()
		cancel()
		if next != maxParallelQueries {
			break
		}
		queries = queries[next:]
	}

	return response, nil
}
```

Reply:
🤣